File: py-appscript/trunk/appscript_2x/lib/aem/typewrappers.py
"""typewrappers -- wrapper classes for AE type and enumeration codes """

######################################################################
# PUBLIC
######################################################################

class AETypeBase(object): # base class; exposed for typechecking purposes
	
	def __init__(self, code):
		# Check arg is a 4-byte code (while ae.newdesc() verifies descriptor type codes okay, it doesn't verify data size so wouldn't catch bad values at packing time):
		if not isinstance(code, basestring):
			raise TypeError('invalid code (not a str object): %r' % code)
		elif len(code) != 4:
			raise ValueError('invalid code (not four bytes long): %r' % code)
		self._code = code
	
	code = property(lambda self:self._code)
	
	def __hash__(self): 
		return hash(self._code)
	
	def __eq__(self, val):
		return val.__class__ == self.__class__ and val.code == self._code
	
	def __ne__(self, val):
		return not self == val
		
	def __repr__(self):
		s = ''
		for c in self._code:
			if 33 < ord(c) < 127 and c != "'":
				s += c
			else:
				s += '\\x%2.2x' % ord(c)
		return "aem.%s('%s')" % (self.__class__.__name__, s)


class AEType(AETypeBase):
	"""An AE type."""


class AEEnum(AETypeBase):
	"""An AE enumeration."""


class AEProp(AETypeBase):
	"""An AE property code."""


class AEKey(AETypeBase):
	"""An AE keyword."""

File: rb-appscript/trunk/src/lib/_aem/typewrappers.rb
#
# rb-appscript
#
# typewrappers -- Ruby classes representing type, enumerator, property and keyword AEDescs
#

module TypeWrappers
	
	class AETypeBase
		
		attr_reader :code
		
		def initialize(code)
			if not(code.is_a?(String) and code.length == 4)
				raise ArgumentError, "Code must be a four-character string: #{code.inspect}"
			end
			@code = code
		end
		
		def hash 
			return @code.hash
		end
		
		def ==(val)
			return (self.equal?(val) or (val.class == self.class and val.code == @code))
		end
		
		alias_method :eql?, :== 
			
		def to_s
			return "AEM::#{self.class::Name}.new(#{@code.inspect})"
		end
		
		def inspect
			return to_s
		end
	end
	
	
	class AEType < AETypeBase
		Name = 'AEType'
	end
	
	
	class AEEnum < AETypeBase
		Name = 'AEEnum'
	end
	
	
	class AEProp < AETypeBase
		Name = 'AEProp'
	end
	
	
	class AEKey < AETypeBase
		Name = 'AEKey'
	end

end